### PR TITLE
Classlib: Event: Fix strum, issue #4405

### LINKED ISSUE #4405: Event: negative strum Bug and Fix
## Steps to reproduce

```supercollider
e = ();
e.freq = [220, 440, 660];
e.strum = -0.25;
e.play();
```

## Expected vs. actual behavior

Duration behaves like this:

220: ------------
440: ..--------..
660: ....----....

when it should do like this instead:

220: ------------....
440: ..------------..
660: ....------------

## Fix

in line 575 negative strum values reverse bndl order but not ids order

```supercollider
if (strum < 0) { bndl = bndl.reverse };
```

fix:

```supercollider
if (strum < 0) { 
	bndl = bndl.reverse;
	ids = ids.reverse
};
```

that's because bndl instantiate nodes in reverse id order and later, in line 588, ids is used straight to free them.

I made a patch but my sc clone was with divergence to upstream in an css help file... I don't know why that happened and didn't solve it properly and after an hour gave up. I'll try to make a pull request again in the next few days.

### PR DESCRIPTION
**Pull request by smrg-lm**
## Purpose and Motivation

This patch Fixes #4405 issue with strum parameter for Event as described there.

## Types of changes

- Bug fix

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] This PR is ready for review

### CODE REVIEW

**Comment by smrg-lm** [comment]
for the sake of completeness, other way to implement it could be to extract the ids from bndl like ````bndl.flop[2]````.

Line 587 (without this patch), could be:

````supercollider
[15 /* \n_set */, bndl.flop[2], \gate, 0].flop,
````

````bndl```` array is a larger data structure however I don't know if ````flop```` is faster than ````reverse```` for this case or how much that is important in this context.

This patch follows implementation semantics used before in that function and ````ids```` variable isn't used afterwards (same as bndl). I just thought that to follow current implementation was better in any case.

**Comment by smrg-lm** [comment]
And, I just saw a possible source of confusion:

Before, line 554:

````~id = ids = Array.fill(bndl.size, { server.nextNodeID });````

should ````~id```` be reassigned? I don't think so because ids are just synth instantiated in reverse temporal order, ````~freq```` wouldn't change either for example.

The thing is that ````ids```` was not a constant in that block, and why is used as ````nil```` in line 537 is something I don't want to ask.

I'm just sharing my reasoning about this.

**Comment by telephon** [comment]
> ````bndl```` array is a larger data structure however I don't know if ````flop```` is faster than ````reverse```` for this case or how much that is important in this context.

reversing an array of a couple of numbers will certainly be faster than a flop of the whole structure. If in doubt, you can check, e.g.:

```supercollider
a = (0..32);
b = [2, a, 5];

bench { a.reverse };
bench { b.flop };
```

> should `~id` be reassigned? 

No, as you say, the order of the variable `ids` is really only the order for the scheduling timing, not the order of addressing.

> The thing is that ids was not a constant in that block, and why is used as nil in line 537 is something I don't want to ask.

I'll answer anyway. It in `nil` because we don't know how many synths (and therefore how many ids) we'll need before multichannel expansion in  https://github.com/supercollider/supercollider/blob/a805ec9d8b0651915a11d85f5c6468b58e2fdf75/SCClassLibrary/Common/Collections/Event.sc#L535-L538

**Comment by smrg-lm** [comment]
Thanks for the review Julian. My first thought about the variable was it was not initialised, "that's wrong", then I thought it's there because clearly marks the place for the ids defined later as a name instead of raw nil. However, after that I decided for me it would have been better to use nil because the next operations in that array slot makes clear that there go ids. I was a little jocular.

**Comment by telephon** [comment]
ah, yes! Thanks for the discussion.

File: SCClassLibrary/Common/Collections/Event.sc [hunk 1]
```diff
@@ -564,8 +564,10 @@ Event : Environment {
 								);
 							}
 						} {
-
-							if (strum < 0) { bndl = bndl.reverse };
+							if (strum < 0) {
+								bndl = bndl.reverse;
+								ids = ids.reverse
```
Comment (snappizz) [inline SCClassLibrary/Common/Collections/Event.sc]: sorry to nitpick but this needs a semicolon:

```suggestion
								ids = ids.reverse;
```

**Comment by patrickdupuis** [comment]
Should this be merged into `3.10` as well?

**Pull request merged**